projects/packages/waf/tests/php/test-waf-runtime.php

```diff
@@ -29,6 +29,140 @@ protected function before() {
 		$this->runtime = new WafRuntime( new WafTransforms(), new WafOperators() );
 	}
 
+	/**
+	 * Test array_flatten
+	 */
+	public function testArrayFlatten() {
+		// get the private array_flatten method
+		$class  = new \ReflectionClass( $this->runtime );
+		$method = $class->getMethod( 'array_flatten' );
+		$method->setAccessible( true );
```

Comment:
I can't say that I particularly like the approach of testing the private methods directly here, for two reasons:

1) It means that our tests will break if we change implementation details of something that is actually private to the class, i.e. rename the method.
2) We don't test the public interface, meaning we don't verify if sending the input with nested array to our runtime actually flattens the arrays / behaves as expected. It could be possible we never fall these private methods but our unit tests were still green (and depending on how you define "unit" that may be fine, but for me it should be the whole class).

Those are certainly philosophical questions though. Other than that, LGTM!

Reply:
I agree with @ArSn here. Unit tests should test the exposed APIs of a unit, not its private implementation details. Iow, test that the unit does the right thing, not that it's implemented in a certain way.

Reply:
Yeah, I mostly agree with that. The issue the other way is that then what you're actually testing is a long method chain rather than just the flatten/normalize, and that ends up being more a test of whether you can construct/mock all the other arguments that you don't actually care about.

Really the issue is that we're using a class as a namespace for a lot of these functions; they only call `$this->{other_function}` and never use `$this->{property}`, suggesting they should be factored out (and would then be testable separately, as desired).